File: Striver/DP/Q25.java
package Striver.DP;

public class Q25{
    // longest common subsequence

    // recursive solution: Time: O(2^n) Space: O(n)
    public static int lcs(String s1, String s2, int n, int m){
        if(n < 0 || m < 0) return 0;
        if(s1.charAt(n) == s2.charAt(m)) return 1 + lcs(s1, s2, n-1, m-1);
        else return Math.max(lcs(s1, s2, n-1, m), lcs(s1, s2, n, m-1));
    }

    // memoization: Time: O(n*m) Space: O(n*m)
    public static int lcs_memo(String s1, String s2, int n, int m, int[][] dp){
        if(n < 0 || m < 0) return 0;
        if(dp[n][m] != -1) return dp[n][m];
        if(s1.charAt(n) == s2.charAt(m)) return dp[n][m] = 1 + lcs_memo(s1, s2, n-1, m-1, dp);
        else return dp[n][m] = Math.max(lcs_memo(s1, s2, n-1, m, dp), lcs_memo(s1, s2, n, m-1, dp));
    }

    // tabulation: Time: O(n*m) Space: O(n*m)
    public static int lcs_tab(String s1, String s2, int n, int m){
        int[][] dp = new int[n+1][m+1];
        for(int i = 0; i <= m; i++) dp[0][i] = 0;
        for(int i = 0; i <= n; i++) dp[i][0] = 0;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s1.charAt(i-1) == s2.charAt(j-1)) dp[i][j] = 1 + dp[i-1][j-1];
                else dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
            }
        }

        return dp[n][m];
    }

    // space optimization: Time: O(n*m) Space: O(n)
    public static int lcs_tab_opt(String s1, String s2, int n, int m){
        int[] dp = new int[m+1];
        int[] curr = new int[m+1];
        for(int i = 0; i <= m; i++) dp[i] = 0;

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s1.charAt(i-1) == s2.charAt(j-1)) curr[j] = 1 + dp[j-1];
                else curr[j] = Math.max(dp[j], curr[j-1]);
            }
            for(int j = 0; j <= m; j++) dp[j] = curr[j];
        }

        return dp[m];
    }

    public static void main(String[] args){
        String s1 = "abcdgh";
        String s2 = "abedfhr";
        int n = s1.length();
        int m = s2.length();
        int[][] dp = new int[n][m];
        for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) dp[i][j] = -1;
        System.out.println(lcs(s1, s2, n-1, m-1));
        System.out.println(lcs_memo(s1, s2, n-1, m-1, dp));
        System.out.println(lcs_tab(s1, s2, n, m));
        System.out.println(lcs_tab_opt(s1, s2, n, m));
    }
}

File: Striver/DP/Q25.py
# longest common subsequence

# recursive solution
def lcs(s1, s2, idx1, idx2):
    if idx1<0 or idx2<0:
        return 0
    
    if s1[idx1] == s2[idx2]:
        return 1 + lcs(s1, s2, idx1-1, idx2-1)
    
    return max(lcs(s1, s2, idx1-1, idx2), lcs(s1, s2, idx1, idx2-1))


# memoization
def lcsMem(s1, s2, idx1, idx2, dp):
    if idx1<0 or idx2<0:
        return 0
    
    if dp[idx1][idx2] != -1:
        return dp[idx1][idx2]
    
    if s1[idx1] == s2[idx2]:
        dp[idx1][idx2] = 1 + lcsMem(s1, s2, idx1-1, idx2-1, dp)
        return dp[idx1][idx2]
    
    dp[idx1][idx2] = max(lcsMem(s1, s2, idx1-1, idx2, dp), lcsMem(s1, s2, idx1, idx2-1, dp))
    return dp[idx1][idx2]

# tabulation
def lcsTab(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [[0] * (n+1) for _ in range(m+1)]

    for i in range(0,m):
        dp[0][i] = 0

    for i in range(0,n):
        dp[i][0] = 0

    for i in range(1,n+1):
        for j in range(1,m+1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = 1 + dp[i-1][j-1]
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])
    
    return dp[n][m]

# space optimization
def lcsTabOpt(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [0] * (m+1)
    curr = [0] * (m+1)

    for i in range(1,m+1):
        dp[i] = 0

    for i in range(1,n+1):
        for j in range(1,m+1):
            if s1[i-1] == s2[j-1]:
                curr[j] = 1 + dp[j-1]
            else:
                curr[j] = max(dp[j], curr[j-1])
        dp = curr
        curr = [0] * (m+1)

    return dp[m]


s1 = "abcdgh"
s2 = "aedfhr"
print(lcs(s1, s2, len(s1)-1, len(s2)-1))
dp = [[-1] * (len(s2)+1) for _ in range(len(s1)+1)]
print(lcsMem(s1, s2, len(s1)-1, len(s2)-1, dp))
print(lcsTab(s1, s2))
print(lcsTabOpt(s1, s2))